pilot/pkg/xds/ecds.go

```diff
@@ -71,5 +114,87 @@ func (e *EcdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w
 			Resource: util.MessageToAny(c),
 		})
 	}
+
 	return resources, model.DefaultXdsLogDetails, nil
 }
+
+func (e *EcdsGenerator) GeneratePullSecrets(proxy *model.Proxy, updatedSecrets map[model.ConfigKey]struct{}, secretResources []SecretResource,
+	secretController credscontroller.Controller, req *model.PushRequest) map[string][]byte {
+	if proxy.VerifiedIdentity == nil {
+		log.Warnf("proxy %s is not authorized to receive secret. Ensure you are connecting over TLS port and are authenticated.", proxy.ID)
+		return nil
+	}
+
+	results := make(map[string][]byte)
+	for _, sr := range secretResources {
+		if updatedSecrets != nil {
+			if _, found := updatedSecrets[model.ConfigKey{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace}]; !found {
+				// This is an incremental update, filter out credscontroller that are not updated.
+				continue
+			}
+		}
+
+		cred, err := secretController.GetDockerCredential(sr.Name, sr.Namespace)
+		if err != nil {
+			log.Warnf("Failed to fetch docker credential %s: %v", sr.ResourceName, err)
+		} else {
+			results[sr.ResourceName] = cred
+		}
+	}
+	return results
+}
+
+func (e *EcdsGenerator) SetCredController(creds credscontroller.MulticlusterController) {
+	e.secretController = creds
+}
+
+func referencedSecrets(proxy *model.Proxy, push *model.PushContext, resourceNames []string) []SecretResource {
+	// The requirement for the Wasm pull secret:
+	// * Wasm pull secrets must be of type `kubernetes.io/dockerconfigjson`.
+	// * Secret are referenced by a WasmPlugin which applies to this proxy.
+	// TODO: we get the WasmPlugins here to get the secrets reference in order to decide whether ECDS push is needed,
+	//       and we will get it again at extension config build. Avoid getting it twice if this becomes a problem.
+	watched := sets.NewSet(resourceNames...)
+	wasmPlugins := push.WasmPlugins(proxy)
+	referencedSecrets := sets.Set{}
+	for _, wps := range wasmPlugins {
+		for _, wp := range wps {
+			if watched.Contains(wp.ResourceName) && wp.ImagePullSecret != "" {
+				referencedSecrets.Insert(wp.ImagePullSecret)
+			}
+		}
+	}
+	var filtered []SecretResource
+	for rn := range referencedSecrets {
+		sr, err := parseSecretName(rn, proxy.Metadata.ClusterID)
+		if err != nil {
+			log.Warnf("Failed to parse secret resource name %v: %v", rn, err)
+			continue
+		}
+		filtered = append(filtered, sr)
+	}
+
+	return filtered
+}
+
+func parseSecretName(resourceName string, proxyCluster cluster.ID) (SecretResource, error) {
```

Comment:
`parseSecretName` decode `resourceName` generated by `toSecretResourceName`, can we put them together or add some comment?

Reply:
Added some comments.